### ISSUE TITLE
AngularJS => Angular migration and ng-annotate

### ISSUE THREAD

**Issue by yjaaidi** [angular/angular-cli#6873]
### Feature Request
```
- [ ] bug report -> please search issues before submitting
- [x] feature request
```

### Desired functionality.
Hi, I'm actually helping some clients to migrate from AngularJS to Angular. The whole AngularJS is already using webpack and typescript.
The hybrid app is actually working with `@angular/cli` but we lost `ng-annotate`.

What would be the best way to add `ng-annotate` to `@angular/cli` without "ejecting" the webpack configuration and customizing it?

Wouldn't it be nice to include ng-annotate in `@angular/cli` for all those who are currently migrating?

For the moment, the workaround I am using is to replace `loader: webpackLoader` with `loaders: ['ng-annotate-loader', webpackLoader]` in an `@angular/cli` fork as you can see here https://github.com/wishtack/angular-cli-ng1/tree/ng-annotate.
WARNING: You should use `commonjs` modules instead of `es2015` (in tsconfig.json) as `ng-annotate` doesn't understand es2015 imports.

Thank you in advance.

Younes

**Comment by filipesilva** [comment]
I'm sorry but this isn't something we're looking at adding. I agree it's a useful tool, and have used it myself, but it's outside of the scope of the CLI.

**Comment by yjaaidi** [comment]
Hi Filipe,

Thank you for your reply.

So, if you are familiar with ng-annotate, you know that it's required by AngularJS for minification, otherwise developers would have to annotate manually which is inacceptable.
As we are all looking forward to help people migrate from AngularJS to Angular, how can they migrate if we don't encourage them by letting the CLI handle this for them.

My clients are migrating their AngularJS apps to Angular and I'm embarrassed to ask them to use my Angular CLI fork in order to benefit from ng-annotate and the CLI.

Isn't `@angular/upgrade` without `ng-annotate` non-sense?

Thank you for your help or any alternative solution.

**Comment by filipesilva** [comment]
Angular CLI is primarily a Angular development tool, not a AngularJS one. Adding extra parts to the build pipeline is not without cost. In fact, it would incur a build and rebuild performance cost on all projects. In this case ejecting is a good alternative. It allows you to cater builds to your specific needs.

**Comment by yjaaidi** [comment]
Thanks for your reply.

It can be optional with something like an `angularJsSupport` flag in
`.angular-cli.json`.

Most of my clients are migrating from AngularJS to Angular and feel a bit
neglected by the CLI.

Asking AngularJS developers who are migrating to Angular to use `ng eject`
sounds quite rude for me. We're not encouraging people to migrate here
while it would be so easy to just add an optional flag to enable
ng-annotate.

Sorry for insisting here. I feel a bit annoying but I think that it's a
major issue concerning Angular adoption by AngularJS developers.

Thanks for your help!

**Comment by kencaron** [comment]
@yjaaidi Thanks for your branch. I've found the lack of this feature pretty glaring, it renders the Angular CLI build chain useless for all apps using ngUpgrade.

Anyway, I've been trying to test it out in a project with `npm install --save-dev github:wishtack/angular-cli#845a1d44d9f0c27b969e932b90b4dc12b2c6bed9` Which _should_ point to the latest commit at your [wishtack/angular-cli@ng-annotate-1.3.1](https://github.com/wishtack/angular-cli/commits/ng-annotate-1.3.1) branch. I've had trouble running this fork of the CLI. If you happen to see this, if you wouldn't mind documenting the steps needed to run it, that'd be very helpful!

**Comment by frederikprijck** [comment]
> In this case ejecting is a good alternative.

I don't think it's a good idea to force everyone to eject who (finally) wants to migrate (using ngUpgrade, which does get promoted at conferences) and have a decent AngularJS setup (I do consider using ngAnnotate a good idea for AngularJS projects).

This would result in all the migrated people moving away from cli (which doesn't sound like something you want, as I've seen that being said multiple times at the gitter/github), moving back will again be a big step to take once all AngularJS code is migrated away. 

----

Not saying this should be part of the CLI. Just saying I do think this will make using ngUpgrade even harder and keep people even further away from upgrading.
 
ngUpgrade has other requirements as well before being able to upgrade, but forcing people to either eject or rewrite the `$inject` array (either manually to using a script) doesn't really sound attractive.

**Comment by yjaaidi** [comment]
Hi @kencaron!
Thank you for your feedback. In fact, it's tricky and due to the monorepo approach, the only way to use a fork for the moment is to release a package with a different name or to follow these instructions https://github.com/angular/angular-cli#development-hints-for-working-on-angular-cli.

**Comment by scipper** [comment]
I've found some kind of hacky solution. Injection all dependencies was not very nice, as our application uses the ngInject annotation in 605 places. 

So I added the ngAnnotation loader to the plugin list of the webpack config of @angular-devkit.

First: my current versions:
"@angular-devkit/build-angular": "~0.11.0",
"@angular/cli": "~7.1.1",
"@angular/compiler-cli": "~7.1.0",
"@angular/language-service": "~7.1.0",
"@angularclass/hmr": "^2.1.3",
"ng-annotate-webpack-plugin": "0.3.0",

Second: open the file `/your-project-root/node_modules/@angular-devkit/build-angular/src/angular-cli-files/models/webpack-configs/browser.js`

Third: add `const ngAnnotatePlugin = require('ng-annotate-webpack-plugin');` to the top and 
`
    extraPlugins.unshift(new ngAnnotatePlugin({
        add: true
    }));
` right before the return.

This way I could make the build process use the ngAnnotate loader and it converts my ngInject annotations to $inject properties.
